---
title: "QUBIC tutorial"
author:
- Yu Zhang
- Qin Ma
classoption: hyperref,
output:
  pdf_document:
    highlight: tango
    latex_engine: xelatex
  html_document: default
  word_document:
    fig_caption: yes
    fig_height: 5
    fig_width: 5
    highlight: tango
    md_extensions: -autolink_bare_uris
bibliography: bibliography.bib
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{QUBIC vignette}
-->

```{r, echo=FALSE}
#Sys.setenv(BIBINPUTS=getwd()); #Set correct path for bibliohraphy file #     citation_package: biblatex

```

Gene expression data is very important in experimental molecular biology [@brazma00], especially for cancer study [@fehrmann15]. The large-scale microarray data and RNA-seq data provide good opportunity to do the gene co-expression analyses and identify co-expressed gene modules; and the effective and efficient algorithms are needed to implement such analysis. Substantial efforts have been made in this field, such as @cheng00, Plaid [@lazzeroni02], Bayesian Biclustering (BCC) [@gu08], among them Cheng and Church and Plaid has the R package implementation. It is worth noting that our in-house biclustering algorithm, QUBIC @li09, is reviewed as one of the best programs in terms of their prediction performance on benchmark datasets. Most importantly, it is reviewed as the best one for large-scale real biological data [@eren12].

Until now, QUBIC has been cited over 90 times (via Google Scholar) and its web server, QServer [@zhou12], was developed in 2012 to facilitate the users without comprehensive computational background [@zhou12]. In the past five years, the cost of RNA-sequencing decreased dramatically, and the amount of gene expression data keeps increasing. Upon requests from users and collaborators, we developed this R package of QUBIC to void submitting large data to a webserver.

The unique features of our R package include (1) updated and more stable back-end resource code (re-written by C++), which has better memory control and is more efficient than the one published in 2009. For an input dataset in Arabidopsis, with 25,698 genes and 208 samples, we observed more than 40% time saving; and (2) comprehensive functions and examples, including discretize function, heatmap drawing and network analysis. 


# Other languages

If R is not your thing, there is also a C version of [`QUBIC`](https://github.com/maqin2001/QUBIC).

# Help 

If you are having trouble with this R package, contact [the maintainer, Yu Zhang](mailto:zy26@jlu.edu.cn). 

# Install and load

Stable version from BioConductor


```r
source("https://bioconductor.org/biocLite.R")
biocLite("QUBIC")
```

<!-- Or development version from GitHub


```r
install.packages("devtools")
devtools::install_github("ropensci/alm")
```
-->
Load `QUBIC`


```{r, message=FALSE}
library("QUBIC")
```

# Example of a random matrix with one embedded bicluster

We can manually create a big bicluster block inside a random matrix:

```{r}
test <- matrix(rnorm(5000),100,50)
test[11:20,11:20] <- rnorm(100,3,0.3)
res<-biclust(test, method = BCQU())
```

We can use `biclust()` function with `BCQU()` as the method parameter. `res` should be the result.

```{r}
summary(res)
```

We can use `summary()` function to show the number, size, rows and columns of clusters in `res`. 

```{r}
show(res)
```

We can use `show()` function to show a brief version. 

```{r}
names(attributes(res))
```

We can see the attributes of `res`, which could use for further processing.

# Example of a random matrix with three diferent embedded biclusters

```{r}
library(QUBIC)
set.seed(1);
# Create a random matrix
test <- matrix(rnorm(10000), 100, 100)
colnames(test) <- paste("gene", 1:100, sep="_")
rownames(test) <- paste("cond", 1:100, sep="_")

# Fill bicluster blocks
t1 <- runif(10, 0.8, 1)
t2 <- runif(10, 0.8, 1) * (-1)
t3 <- runif(10, 0.8, 1) * sample(c(-1, 1), 10, replace = TRUE)
test[11:20, 11:20] <- t(rep(t1, 10) * rnorm(100, 3, 0.3))
test[31:40, 31:40] <- t(rep(t2, 10) * rnorm(100, 3, 0.3))
test[51:60, 51:60] <- t(rep(t3, 10) * rnorm(100, 3, 0.3))

# QUBIC
res <- biclust(test, method = BCQU())
summary(res)

# Show heatmap
hmcols <-  colorRampPalette(rev(c("#D73027", "#FC8D59", "#FEE090", "#FFFFBF", "#E0F3F8", "#91BFDB", "#4575B4")))(100)
# Specify colors

par(mar=c(1,5,5,1)+0.1)
heatmapBC(x = test, bicResult = res, number = c(1, 2, 3), local = FALSE,
          col = hmcols, xaxt="n", xlab="", ylab="")

mycolnames1 <- colnames(test)[res@RowxNumber[,1]]
mycolnames2 <- colnames(test)[res@RowxNumber[,2]]
mycolnames3 <- colnames(test)[res@RowxNumber[,3]]
mycolnames <- c(mycolnames1, mycolnames2, mycolnames3)

axis(2, at = length(mycolnames):1, labels = mycolnames, cex.axis = 1.0, las=1)
myrownames1 <- rownames(test)[res@NumberxCol[1,]]
myrownames2 <- rownames(test)[res@NumberxCol[2,]]
myrownames3 <- rownames(test)[res@NumberxCol[3,]]
myrownames <- c(myrownames1, myrownames2, myrownames3)

axis(3, at = 1:length(myrownames), labels = myrownames, las=2)

```

Functions
======================

There are three functions provide by QUBIC package. ```BCQU()``` makes
following example showed biclusters in Yeast.

```{r, cache=TRUE}
library(QUBIC);
data(BicatYeast)
x <- BicatYeast
system.time(res <- biclust(x, method = BCQU()))

summary(res)
```


We can also draw heat maps.


```{r}
# Draw heatmap of the first cluster
drawHeatmap(x, res, 1)

# Draw overlapping heatmap
heatmapBC(x = x, bicResult = res, number = c(1, 2), local = TRUE)
net <- qunetwork(x, res, number = c(1, 2), group = c(1, 2), method = "spearman")
```

```{r, results='hide'}
# Output overlapping heatmap XML, could be used in other software such as
# Cytoscape, Biomax or JNets
qunet2xml(net, minimum = 0.6, color = cbind(rainbow(length(net[[2]]) - 1), "gray"))
# We can use Cytoscape, Biomax or JNets open file named tempnetworkresult.gr
```

# References
