---
title: "Supplementary Data"
author:
  - name: Author One
    affiliation: Affiliation
    address:
    - line 1
    - line 2
    email:  author1@work
  - name: Author Two
    affiliation: Affiliation
    address:
    - line 1
    - line 2
    email:  author2@work
date: "15 June 2015"
output:
  pdf_document:
    fig_caption: yes
    latex_engine: xelatex
    number_sections: yes
    toc_depth: 3
  html_document:
    toc: yes
  word_document:
    md_extensions: -autolink_bare_uris
    reference_docx: Bioinformatics.docx
bibliography: ../vignettes/bibliography.bib
csl: bioinformatics.csl
preamble: |
  % Any extra latex you need in the preamble
abstract: |
  An abstract of less than 150 words.   The abstract. **Summary**:  **Availability**:<!---张禹。请参考这个文章的摘要把摘要中的summary和availabillity完成。--> **Contact**: qin.ma@sdstate.edu **Supplementary information:** Supplementary data are available at *Bioinformatics* online.
---
```
opts_knit$set('upload.fun' = function(x) sub("\\.[[:alnum:]]*$", "", x) ) 
```
<!--pandoc
format: latex
latex-engine: xelatex
default-image-extension: pdf

format: html
default-image-extension: png
-->
---

Yu Zhang^1^, Qin Ma^2,\#^

^1^College of Computer Science and Technology, Jilin University,
Changchun, 130012, China

^2^Department of Plant Science, South Dakota State University,
Brookings, SD, 57007, USA

^\#^ Corresponding author: qin.ma@sdstate.edu

**Running title:** An R package for biclustering of gene expression data

Abstract
========


Introduction
============

QUBIC is a qualitative biclustering algorithm for analyses of gene
expression data authored by @li09. We now give the new
implementation in R.

1.  Gene expression data is very important in experimental molecular biology [@brazma00], especially for cancer study [@fehrmann15].<!---Try to find some description in QUBIC paper and other related papers. 注意引用相关文献。最好有些最新的。-->

2.  Large-scale microarray data and RNA-seq data provide good
    opportunity to do gene co-expression analyses; meanwhile the
    effective and efficient algorithms are needed to implement such
    analysis.

3.  Substantial efforts have been made in this field, such as @cheng00, Plaid [@lazzeroni02], Bayesian Biclustering (BCC) [@gu08]<!---列举出主要的几个双聚类软件。具体的可以参见这个review：http://bib.oxfordjournals.org/content/early/2012/07/06/bib.bbs032.full-->, among them Cheng and Church and Plaid has the R package implementation. It is worth
    noting that our in-house algorithm, QUBIC, is reviewed as one of the
    best programs in terms of their prediction performance on benchmark
    datasets. Most importantly, it is the best one regarding real
    biological data [@eren12].

4.  Currently, QUBIC is already cited for 88 times (via Google Scholar) and in 2012 we also
    developed a web server, QServer, to facilitate the users without
    comprehensive computational background [@zhou12]. As the
    scale of gene expression data keeps increasing, we developed this R
    package of QUBIC to void submitting large data to a webserver.

5.  The unique features of our R package:

    a.  Updated and more stable back-end resource code, written by C++,
        which has better memory control and is more efficient than the
        one published in 2009. 
        For a input with 50M (arabidopsis-leaf), we observed more than 40% time saving.
        <!--- you may want to explain why it is more efficient. 也就是你对源程序的关键改动和具体提高的程度百分比。 -->

    b.  Comprehensive functions and examples, including discretize function, heatmap drawing and network analysis.<!---please enrich this part,这里要突出我们所有的功能，譬如离散化，图形化，网络分析以及围绕bicluster的各种小功能。-->

Implementation
==============

QUBIC (QUBIC 1.0) original written in C using POSIX apis. The
portability is limited. Some memory leak will be occur if trying to call
the primarily functions more than once.

We did refactoring and rewrote most part of QUBIC 1.0 in C++ and try to keep
the result is the same. We changed most data structures to avoid memory leak. 
Most of C pointers are replaced with std containers. 
We rewrote and optimized core functions with the result of profiling,
<!--- 这个可以更加具体，到底怎样改的，譬如更换了fibunaqie 堆等等。同时提一下这个版本非常方便进一步的功能开发和更新。-->

The performance is increased. We found average half time saving. (Windows 8 x64, Memory 8G, Intel Core i7-3770 3.4G) Large files will find more savings.

Input Data          Before (s)      After (s)
----------------  ------------   ------------
Arabidopsis-leaf         24.29          16.41
E_coli_v4               181.38          89.58
BRCA_tumor              944.25         481.60

Table:  Performance increased.

```{r}
data <- structure(list(E = c(24.29, 16.41),
                       A = c(181.38, 89.58),
                       B = c(944.25, 481.60)),
                  .Names = c("E_coli\n(32% saved)", 
                             "arabidopsis\n(51% saved)", 
                             "BRCA_tumor\n(49% saved)"),
                  class = "data.frame", row.names = c(NA, -2L))
attach(data)
print(data)
colours <- rainbow(2)
midpoints <- barplot(as.matrix(data), ylab = "Time (seconds)", 
                     cex.lab = 1.5, cex.main = 1.4, beside=TRUE, col=colours)
legend("topleft", c("QUBIC 1.0","QUBIC R package"), cex=1.3, bty="n", fill=colours)
```

<!--- 要有具体的百分比。这里建议在附件中加上一个图或者表格，在几套不同的数据上比较一下原来版本和目前版本的时间。
同时可以引用QUBIC的原文，在附件中我当时比较了时间，QUBIC是最快的。所以这里你只要说明Rpackage比QUBIC更快就好了。-->

The R package could be working with biclust<!---除了这个，我们还要提我们可以和cytoscape结合做后续的网络化分析。 --> written by @kaiser09.
The R package also provide output which could use as input for software such as Cytoscape [@smoot11].

We also give examples of heat map and overlapping networks.

Functions and examples
======================

There are three functions provide by QUBIC package. ```BCQU()``` makes
following...
```{r, cache=TRUE}
library(QUBIC);
tmpdir <- tempdir()
rdata_file <- paste(tmpdir, "/E_coli_v4_Build_6.RData", sep = "")
if (file.exists(rdata_file)) {
  load(rdata_file);
} else {
  library(tools);
  file <- paste(tmpdir, "/E_coli_v4_Build_6_chips907probes4297.tab",
                sep = "")
  if(!file.exists(file) || (file.info(file)$size != 30822861) ||
     (as.vector(md5sum(file)) != "24702063500c9c88fec54c7ce65cd1da")) {
    url <- "http://m3d.mssm.edu/norm/E_coli_v4_Build_6.tar.gz"
    size <- 117091420
    md5 <- "8092cb77106d63768d62b24bff1a48d9"
    down <- function(url, size, md5){
      library(tools)
      file <- paste(tmpdir, '/', basename(url), sep = "")
      if(file.exists(file)) {
        if((file.info(file)$size != size) || (as.vector(md5sum(file)) != md5)) {
          file.remove(file)
        } else {
          return(file)
        }
      }
      download.file(url, file)
      return(file)
    }
    tar_file <- down(url, size, md5)
    untar(tar_file, compressed = 'gzip')
    file.rename("E_coli_v4_Build_6/E_coli_v4_Build_6_chips907probes4297.tab", file)
    unlink("E_coli_v4_Build_6", recursive=TRUE)
  }
  x <- as.matrix(read.table(file, row.names = 1, header=T))
  system.time(res <- biclust::biclust(x, method = BCQU()))
  save(x, res, file= rdata_file)
}

summary(res)
```


We can also draw heat maps.


```{r}
# Draw heatmap of the first cluster
biclust::drawHeatmap(x, res, 1)

# Draw overlapping heatmap
biclust::heatmapBC(x = x, bicResult = res, number = c(1, 2), local = TRUE)
net <- qunetwork(x, res, number = c(1, 2), group = c(1, 2), method = "spearman")
```

```{r, results='hide'}
# Output overlapping heatmap XML, could be used in other software such as
# Cytoscape, Biomax or JNets
sink(paste(tmpdir, "/tempnetworkresult.gr", sep = ""))
qunet2xml(net, minimum = 0.6, color = cbind(rainbow(length(net[[2]]) - 1), "gray"))
sink()
# We can use Cytoscape, Biomax or JNets open file named tempnetworkresult.gr
```
![Output Picture in Cytoscape](Cytoscape)

Conclusions
===========

QUBIC is a well-known qualitative biclustering algorithm. QUBIC package
should serve and save time for bioinformatics scientists.

Acknowledgements
================

This work was supported by xxx<!---如果你那边有funding，可以加在这里。-->.

References
==========
